Fix condition declarations and assignments in for loops

gojakuch opened the pull request:
These changes fix the differentiation of variable declarations in for loop conditions that used to result into wrong derivatives. The commit also tackles the problem of having an assignment operator that affects the derivative in the for-loop condition, as well as adds support for logical operators in for-loops and allows to combine assignments with them.

Fixes: #273

vgvassilev commented on lib/Differentiator/BaseForwardModeVisitor.cpp:
```diff
@@ -668,18 +668,39 @@ StmtDiff BaseForwardModeVisitor::VisitForStmt(const ForStmt* FS) {
+  StmtDiff condDiff = Clone(FS->getCond());
+  if (Expr* cond =
+          condDiff
+              .getExpr()) { // this adds support for assignments in conditions
```
Can you move the comments on a separate line before the statement that we are commenting on? That will help clang-format avoid the weird indentations...

vgvassilev commented on lib/Differentiator/BaseForwardModeVisitor.cpp:
```diff
@@ -668,18 +668,39 @@ StmtDiff BaseForwardModeVisitor::VisitForStmt(const ForStmt* FS) {
+      if (condBO->isAssignmentOp())
+        condDiff = Visit(new (m_Context) ParenExpr(
+            noLoc, noLoc,
+            cond)); // if it's an assignment operator we wrap it back into
```
Likewise.

parth-07 commented:
The fix does not work for all cases. In the case when the for-loop condition fails, we should run the derivative statement for the for-loop condition, but we do not. This leads to incorrect derivative in some cases. 

Reproducible example:

```cpp
#include "clad/Differentiator/Differentiator.h"
#include <iostream>
#define show(x) std::cout << #x << ": " << x << "\n";

double fn(double u, double v) {
    double res = 0;
    for (; (res = u * v) && false ;) {}
    return res;
}

int main() {
    auto fn_diff = clad::differentiate(fn, "u");
    show(fn_diff.execute(3, 5));
}
```
Output:
fn_diff.execute(3, 5): 0

Expected Output:
fn_diff.execute(3, 5): 5

parth-07 reviewed (changes requested):
Please see my earlier comment for the case which produces wrong output with this solution.

vgvassilev commented on lib/Differentiator/BaseForwardModeVisitor.cpp:
```diff
@@ -668,18 +668,37 @@ StmtDiff BaseForwardModeVisitor::VisitForStmt(const ForStmt* FS) {
   StmtDiff initDiff = init ? Visit(init) : StmtDiff{};
   addToCurrentBlock(initDiff.getStmt_dx());
+
+  // declaration in the condition (if any) needs to be differentiated
```
```suggestion
  // Declaration in the condition (if any) needs to be differentiated.
```

Our comments are full sentences starting with a capital letter and ending with punctuation. Can you fix the rest of the comments, too?

gojakuch replied:
okay, I will fix this together with what Parth asked for.

parth-07 commented:
It seems that we do not support logical operators in a for-loop conditions. Hence, support for conditions such as this `(res = u * v) && false` can be added in a separate pull-request. However, the below example should work with this pull request:

```cpp
#include "clad/Differentiator/Differentiator.h"
#include <iostream>
#define show(x) std::cout << #x << ": " << x << "\n";

double fn(double u, double v) {
    double res = u * v;
    for (; (res = 0) ;) {}
    return res;
}

int main() {
    auto fn_diff = clad::differentiate(fn, "u");
    show(fn_diff.execute(3, 5));
}
```

**Output:** 
fn_diff.execute(3, 5): 5

**Expected output:** 
fn_diff.execute(3, 5): 0

gojakuch commented:
> It seems that we do not support logical operators in a for-loop conditions.

I have already started working on this because of your initial suggestion. Adding this as a separate PR would require rewriting the changes made in this PR once again. I can just change the commit and PR messages to also include this once I'm done.

parth-07 commented on lib/Differentiator/BaseForwardModeVisitor.cpp:
```diff
@@ -668,18 +668,50 @@ StmtDiff BaseForwardModeVisitor::VisitForStmt(const ForStmt* FS) {
+
+  // Declaration in the condition (if any) needs to be differentiated.
+  if (VarDecl* condVarDecl = FS->getConditionVariable()) {
+    // FIXME: here DifferentiateVarDecl is only used to clone the declaration
```
What other ways did you try before DifferentiateVarDecl?

gojakuch replied:
I have found another way, I'll push it soon.

parth-07 commented on lib/Differentiator/BaseForwardModeVisitor.cpp:
```diff
@@ -668,18 +668,50 @@ StmtDiff BaseForwardModeVisitor::VisitForStmt(const ForStmt* FS) {
+  // Condition differentiation.
+  // This adds support for assignments in conditions.
+  if (cond) {
+    while (CastExpr* condCast = dyn_cast<CastExpr>(cond))
```
Can these two `while` loops be replaced by: `Expr::IgnoreParenImpCasts` function?

gojakuch replied:
yes, thank you, I didn't know this exists

parth-07 reviewed (changes requested):
It seems that currently for-loop condition is only differentiated if it's a binary operator. This additional check seems unnecessary and is incorrect. You should try to directly call `Visit(...)` function on the condition expression.

The below code gives an incorrect derivative due to the condition expression only being differentiated if it's a binary operator:

```cpp
#include "clad/Differentiator/Differentiator.h"
#include <iostream>
#define show(x) std::cout << #x << ": " << x << "\n";

double fn(double u, double v) {
    double res = 0;
    for (; !(res = u * v) ;) {}
    return res;
}

int main() {
    auto fn_diff = clad::differentiate(fn, "u");
    show(fn(3, 5));
    show(fn_diff.execute(3, 5));
}
```

**Output:**
fn_diff.execute(3, 5): 0

**Expected Output:**
fn_diff.execute(3, 5): 5

gojakuch commented:
> It seems that currently for-loop condition is only differentiated if it's a binary operator. This additional check seems unnecessary and is incorrect. You should try to directly call Visit(...) function on the condition expression.

As for this, I enabled support for unary operators, added support for visiting the logical not in general. I also allowed comparison operators, which are probably the most common type of expression in for-loop conditions, and made the binary operator visitor work with them. I understand that it is a bit ugly to artificially limit the differentiation of cond, however, Clad doesn't support many other operators and if we simply visit anything that is inside cond, some tests start to crash because of complicated reasons related to trying to visit an unsupported statement (that is my guess, I spent some time trying to understand what's going on but I didn't understand much). I don't think we should add general support for other operators in this PR, so I added a fixme message to add support for other expressions in cond. this PR already covers a lot of things that were not possible or were differentiated incorrectly before, so I think visiting only these supported statements could be a common ground and a solution so far.

speaking of crashing checks, as you can see, there are some checks failing for OSX now, specifically the ForwardMode/UserDefinedTypes.C test. which is exactly what started crashing on my machine after I had tried to visit every for-loop condition regardless of its type. these errors here are different though and I don't get them on my machine with the current PR code. I'm not even sure how they are related to for loops. I would appreciate if someone more experienced took a look at these at least, maybe I could make visiting all the conds work with similar fixes then.

parth-07 commented:
@gojakuch You are right. We are enabling a lot of different things in this pull request. It is my bad. I thought some features (binary operator, unary operator) were supported before, but they were not. Can we please limit the scope of this pull request? Let's handle the unary operator and general support for visiting conditions in a separate pull request. 

Can you please revert the latest changes and bring the pull request to the point when you enabled binary operator support and add FIXME in the code regarding general support is missing?

vgvassilev commented:
In addition can you open issues for all things that this pull request discovered are not supported?

gojakuch commented:
> Can you please revert the latest changes and bring the pull request to the point when you enabled binary operator support and add FIXME in the code regarding general support is missing?

done in the last force-push

parth-07 commented on lib/Differentiator/BaseForwardModeVisitor.cpp:
```diff
@@ -751,6 +783,15 @@ StmtDiff BaseForwardModeVisitor::VisitReturnStmt(const ReturnStmt* RS) {
     return nullptr;
 
   StmtDiff retValDiff = Visit(RS->getRetValue());
+  // llvm::errs() << "\n\n\n\n\nretdx: \n";
```
Please remove the debugging statements.

vgvassilev commented on lib/Differentiator/BaseForwardModeVisitor.cpp:
```diff
@@ -668,18 +668,53 @@ StmtDiff BaseForwardModeVisitor::VisitForStmt(const ForStmt* FS) {
+                     nullptr, condVarDecl->getInitStyle());
+    VarDecl* _d_condVarDecl = BuildVarDecl(
+        condVarClone->getType(), "_d_" + condVarClone->getNameAsString(),
+        nullptr, condVarClone->isDirectInit(), nullptr,
```
```suggestion
        /*Init=*/nullptr, condVarClone->isDirectInit(), nullptr,
```
We comment in the the arguments which are unclear what parameters refer to.

vgvassilev commented on lib/Differentiator/BaseForwardModeVisitor.cpp:
```diff
@@ -668,18 +668,53 @@ StmtDiff BaseForwardModeVisitor::VisitForStmt(const ForStmt* FS) {
+    // the declaration. The declaration itself is thrown before the for-loop
+    // without any init value. The fictional condition is then differentiated as
+    // normal condition (see below).
+    cond =
+        BuildOp(BO_Assign, BuildDeclRef(condVarClone), condVarClone->getInit());
```
I'd move the DeclRefExpr from `BuildDeclRef(condVarClone)` into a separate variable. I thin we need to clone the initializer expression of `condVarClone` if we want to use it separately. However, in this situation it is fine because on the next line we remove it. The code needs to become more clear about that. Maybe something like:

CondInit = condVarClone->getInit();
condVarClone->setInit(nullptr);
...

vgvassilev commented on lib/Differentiator/BaseForwardModeVisitor.cpp:
```diff
@@ -668,18 +668,53 @@ StmtDiff BaseForwardModeVisitor::VisitForStmt(const ForStmt* FS) {
+  // Condition differentiation.
+  // This adds support for assignments in conditions.
+  if (cond) {
+    cond = cond->IgnoreParenImpCasts();
```
Shouldn't we strip the implict casts of the init expression when forming it in the previous section?

gojakuch replied:
not always, because if there is no declaration in this for this section will just serve the purpose of differentiating the actual cond from the source code, not the one generated above. in this case it won't be unwrapped and could contain impcasts.

gojakuch replied:
sorry, I didn't understand your question, but this line `cond = cond->IgnoreParenImpCasts();` is needed (at least for now). this way we can check if the cond expression is supported or not, specifically, if there's an assignment operator like this:
```C++
double res;
for(;(r = x);){}
```
this cond is wrapped in implicit casts to bool and parentheses. once we can safely call `Visit(cond)` regardless of its type (that is, once there's support for almost most expressions as for-loop conditions), this line can be removed. but since now we need to check if that is an assignment operator, this is needed.

vgvassilev replied:
The way that it is written assumes that the line `Expr* cond = condDiff.getExpr();` may have implicit casts. Is that the case? If it is not my proposal was to move `cond->IgnoreParenImpCasts()` operation in the prior if-stmt.

vgvassilev commented on lib/Differentiator/BaseForwardModeVisitor.cpp:
```diff
@@ -668,18 +668,53 @@ StmtDiff BaseForwardModeVisitor::VisitForStmt(const ForStmt* FS) {
+    BinaryOperator* condBO = dyn_cast<BinaryOperator>(cond);
+    if (condBO && (condBO->isLogicalOp() || condBO->isAssignmentOp())) {
+      condDiff = Visit(cond);
+      cond = (condDiff.getExpr_dx() && !isUnusedResult(condDiff.getExpr_dx()))
+                 ? BuildOp(BO_Comma, BuildParens(condDiff.getExpr_dx()),
+                           BuildParens(condDiff.getExpr()))
+                 : condDiff.getExpr();
```
That's unreadable. We should probably expand the ternary if.

gojakuch replied:
I agree

vgvassilev commented on lib/Differentiator/BaseForwardModeVisitor.cpp:
```diff
@@ -714,26 +749,23 @@ StmtDiff BaseForwardModeVisitor::VisitForStmt(const ForStmt* FS) {
+  StmtDiff bodyVisited = Visit(body);
+  for (Stmt* S : bodyVisited.getBothStmts())
+    addToCurrentBlock(S);
+  CompoundStmt* bodyResultCmpd = endBlock();
+  if (bodyResultCmpd->size() == 1)
+    bodyResult = bodyResultCmpd->body_front();
+  else
+    bodyResult = bodyResultCmpd;
```
Can't we use `unwrapIfSingleStmt`?

gojakuch replied:
I think I tried but I don't remember why I didn't put it in the PR. maybe I just accidentally reverted it together with something that didn't work out well. in any case, as far as I remember this part of code was already there, I just removed an unnecessary if condition it was wrapped into. I'll try `unwrapIfSingleStmt` again rn.

vgvassilev replied:
Let's move it in a common place and use it.

vgvassilev commented on lib/Differentiator/BaseForwardModeVisitor.cpp:
```diff
@@ -714,26 +749,23 @@ StmtDiff BaseForwardModeVisitor::VisitForStmt(const ForStmt* FS) {
 
   Stmt* forStmtDiff =
-      new (m_Context) ForStmt(m_Context, initDiff.getStmt(), cond, condVarClone,
+      new (m_Context) ForStmt(m_Context, initDiff.getStmt(), cond, nullptr,
```
```suggestion
      new (m_Context) ForStmt(m_Context, initDiff.getStmt(), cond, /*condVar=*/nullptr,
```

vgvassilev commented on lib/Differentiator/BaseForwardModeVisitor.cpp:
```diff
@@ -1366,6 +1407,26 @@ BaseForwardModeVisitor::VisitBinaryOperator(const BinaryOperator* BinOp) {
+    // For (A && B) return ((dA, A) && (dB, B)) to ensure correct evaluation and
+    // correct derivative execution. dLL = (dL, L)
+    Expr* dLL =
+        BuildParens(((Ldiff.getExpr_dx() && !isUnusedResult(Ldiff.getExpr_dx()))
+                         ? BuildOp(BO_Comma, BuildParens(Ldiff.getExpr_dx()),
+                                   BuildParens(Ldiff.getExpr()))
+                         : Ldiff.getExpr()));
```
That is unreadable.

vgvassilev commented on lib/Differentiator/BaseForwardModeVisitor.cpp:
```diff
@@ -1366,6 +1407,26 @@ BaseForwardModeVisitor::VisitBinaryOperator(const BinaryOperator* BinOp) {
+                         : Ldiff.getExpr()));
+    // dRR = (dR, R)
+    Expr* dRR =
+        BuildParens(((Rdiff.getExpr_dx() && !isUnusedResult(Rdiff.getExpr_dx()))
+                         ? BuildOp(BO_Comma, BuildParens(Rdiff.getExpr_dx()),
+                                   BuildParens(Rdiff.getExpr()))
+                         : Rdiff.getExpr()));
```
Likewise. You probably can create some lambda function to avoid duplicating this code in a few places...

vgvassilev commented on lib/Differentiator/BaseForwardModeVisitor.cpp:
```diff
@@ -668,18 +668,53 @@ StmtDiff BaseForwardModeVisitor::VisitForStmt(const ForStmt* FS) {
+        nullptr, condVarClone->isDirectInit(), nullptr,
+        condVarClone->getInitStyle());
+    m_Variables.emplace(condVarClone, BuildDeclRef(_d_condVarDecl));
+    // Here we create a fictional cond that is equal to the assignment used in
+    // the declaration. The declaration itself is thrown before the for-loop
+    // without any init value. The fictional condition is then differentiated as
+    // normal condition (see below).
```
Probably better to give an example with code, too.

vgvassilev commented on lib/Differentiator/BaseForwardModeVisitor.cpp:
```diff
@@ -1366,6 +1405,26 @@ BaseForwardModeVisitor::VisitBinaryOperator(const BinaryOperator* BinOp) {
+      if (Xdiff.getExpr_dx() && !isUnusedResult(Xdiff.getExpr_dx()))
+        return BuildParens(BuildOp(BO_Comma, BuildParens(Xdiff.getExpr_dx()),
+                                   BuildParens(Xdiff.getExpr())));
+      else
```
No else after a return statement.

vgvassilev commented on include/clad/Differentiator/CladUtils.h:
```diff
@@ -27,6 +27,10 @@ namespace clad {
 
+    // Unwraps S to a single statement if it's a compound statement only
+    // containing 1 statement.
+    clang::Stmt* unwrapIfSingleStmt(clang::Stmt* S);
```
That operation should be in a separate commit.

vgvassilev commented on lib/Differentiator/BaseForwardModeVisitor.cpp:
```diff
@@ -1366,6 +1405,25 @@ BaseForwardModeVisitor::VisitBinaryOperator(const BinaryOperator* BinOp) {
+      if (Xdiff.getExpr_dx() && !isUnusedResult(Xdiff.getExpr_dx()))
+        return BuildParens(BuildOp(BO_Comma, BuildParens(Xdiff.getExpr_dx()),
+                                   BuildParens(Xdiff.getExpr())));
+      return BuildParens(Xdiff.getExpr());
```
We should figure out how to test this branch.

gojakuch replied:
what do you mean, like add unit tests for this? this code is executed in the already existing tests if that's what you mean

vgvassilev replied:
Codecov claims this code path is not tested. If that’s true (which is almost always the case) we should add a test.

parth-07 commented on lib/Differentiator/BaseForwardModeVisitor.cpp:
```diff
@@ -668,18 +668,64 @@ StmtDiff BaseForwardModeVisitor::VisitForStmt(const ForStmt* FS) {
+    VarDecl* condVarClone =
+        BuildVarDecl(condVarDecl->getType(), condVarDecl->getNameAsString(),
+                     Clone(condVarDecl->getInit()), condVarDecl->isDirectInit(),
+                     nullptr, condVarDecl->getInitStyle());
```
We are repeating the code of `DifferentiateVarDecl` here just because we have to take the init as `null`. Can you reuse `DifferentiateVarDecl` here by adding an extra argument such as `ignoreInit` to `DifferentiateVarDecl` function?

gojakuch replied:
that seems reasonable. I've implemented this. the thing is though, I needed to create a separate 2-argument method that is called from the 1-argument overload because some other classes specifically override the old single-argument method and this would be messed up by adding an optional second argument.

vgvassilev reviewed (approved):
LGTM! Let's wait for @parth-07's review.

parth-07 reviewed (approved):
Looks good.

vgvassilev commented:
@gojakuch, can you fix the compilation error we got here?
